File: test_regress/t/t_dist_untracked.pl
#!/usr/bin/perl
if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2003 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

scenarios(dist => 1);

my $root = "..";
my $Debug;

if (!-r "$root/.git") {
    skip("Not in a git repository");
} else {
    ### Must trim output before and after our file list
    my %warns;
    my $prefix;
    my $summary;
    {
        my $status = `cd $root && git ls-files -o --exclude-standard`;
        print "ST $status\n" if $Debug;
        foreach my $file (sort split /\n/, $status) {
            next if $file =~ /nodist/;
            if (_has_tabs("$root/$file")) {
                $warns{$file} = "File not in git or .gitignore (with tabs): $file";
                $summary = "Files untracked in git or .gitignore (with tabs):"
            } else {
                $warns{$file} = "File not in git or .gitignore: $file";
                $summary ||= "Files untracked in git or .gitignore:"
            }
        }
    }
    if (keys %warns) {
        # First warning lists everything as that's shown in the driver summary
        error($summary." ",join(' ',sort keys %warns));
        foreach my $file (sort keys %warns) {
            error($warns{$file});
        }
    }
}

sub _has_tabs {
    my $filename = shift;
    my $contents = file_contents($filename);
    if ($filename =~ /\.out$/) {
        # Ignore golden files
    } elsif ($contents =~ /[\001\002\003\004\005\006]/) {
        # Ignore binrary files
    } elsif ($contents =~ /\t/) {
        return 1;
    }
    return 0;
}

ok(1);
1;

File: test_regress/t/t_dist_untracked.py
#!/usr/bin/env python3
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2024 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

import vltest_bootstrap

test.scenarios('dist')

root = ".."

if not os.path.exists(root + "/.git"):
    test.skip("Not in a git repository")

### Must trim output before and after our file list
warns = {}
summary = None

status = test.run_capture("cd " + root + " && git ls-files -o --exclude-standard")

if test.verbose:
    print("-ST " + status)
for filename in sorted(status.split()):
    if re.search('nodist', filename):
        continue
    warns[filename] = "File not in git or .gitignore: " + filename
    summary = "Files untracked in git or .gitignore:"

if summary:
    # First warning lists everything as that's shown in the driver summary
    test.error(summary + " " + ' '.join(sorted(warns.keys())))
    for filename in sorted(warns.keys()):
        test.error(warns[filename])

test.passes()
